ReactDOM.useEvent: wire to event system to the hook

trueadm opened the pull request:
This PR wires up the `useEvent` hook to the ReactDOM renderer. This is the first PR that actually allows you to use `useEvent` properly with ReactDOM.

This PR only adds support for `Element` target nodes that are managed by React are now supported. `document` and `window` targets are not supported just yet (that will come in another PR). I've also added quite a few tests, with more to come.

Lastly, we have to add some additional logic into four places:

- During `executeDispatchesInOrder` we need to check the instance does not match the previous instance when validating `stopPropagation` logic. The reason for this is because it's now possible to have more than one listener per instance (compared to on* prop events).
- During `beforeRemoveInstance` we need to do some extra cleanup when a node is removed has a `useEvent` listener wired to it. If we don't do this, we'll leak the DOM node in the listener (because the listener is referenced in `mountEventListener`'s Map).
- During `accumulateTwoPhaseListeners` we need to also check if the instance has `useEvent` listeners that match the `type` and `capture` flags on the listener.

Still TODO:
- window target support
- document target support
- passive/active event listener upgrading (https://github.com/facebook/react/pull/18308)
- experimental internal scope support

threepointone commented on packages/react-dom/src/events/__tests__/DOMModernPluginEventSystem-test.internal.js:
```diff
@@ -1074,5 +1074,527 @@ describe('DOMModernPluginEventSystem', () => {
+
+    it('should correctly handle many nested target listeners', () => {
+      const buttonRef = React.createRef();
+      const targetListerner1 = jest.fn();
```
```suggestion
      const targetListener1 = jest.fn();
```
(and the ones after this one too)

trueadm commented on packages/react-dom/src/client/ReactDOMHostConfig.js:
```diff
@@ -529,6 +530,22 @@ export function beforeRemoveInstance(
+    // that the hook is attached to gets unmounted.
+    const listenersSet = getListenersFromTarget(instance);
+    if (listenersSet !== null) {
+      const listeners = Array.from(listenersSet);
```
It's far slower in terms of performance.

trueadm commented:
@threepointone Thanks for the review. I addressed all your feedback. In terms of using `createRoot`, there is a test that does that, but it shouldn't really affect things. I'll add more tests that use it in a follow up though.
